File: Master/Year 1/Semester 1/AMSS/Proiect/app/src/main/java/ro/robertto/rescuepets/presentation/adapter/MessagesAdapter.java
package ro.robertto.rescuepets.presentation.adapter;

import android.app.AlertDialog;
import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.database.FirebaseDatabase;

import java.util.ArrayList;

import ro.robertto.rescuepets.R;
import ro.robertto.rescuepets.databinding.DeleteLayoutBinding;
import ro.robertto.rescuepets.databinding.ReceiveMsgBinding;
import ro.robertto.rescuepets.databinding.SendMsgBinding;
import ro.robertto.rescuepets.domain.RescuePetsRemoteRepository;
import ro.robertto.rescuepets.presentation.model.Message;

public class MessagesAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {
    private Context context;
    private ArrayList<Message> messages;
    private String senderRoom;
    private String receiverRoom;
    private final int itemSent = 1;
    private final int itemReceive = 2;

    public MessagesAdapter(Context context, ArrayList<Message> messages, String senderRoom, String receiverRoom) {
        this.context = context;
        this.messages = messages;
        this.senderRoom = senderRoom;
        this.receiverRoom = receiverRoom;
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        if (viewType == itemSent) {
            View view = LayoutInflater.from(context).inflate(R.layout.send_msg, parent, false);
            return new SentMsgHolder(view);
        } else {
            View view = LayoutInflater.from(context).inflate(R.layout.receive_msg, parent, false);
            return new ReceiveMsgHolder(view);
        }
    }

    @Override
    public int getItemViewType(int position) {
        Message message = messages.get(position);
        if (FirebaseAuth.getInstance().getUid().equals(message.getSenderId())) {
            return itemSent;
        } else {
            return itemReceive;
        }
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
        Message message = messages.get(position);

        if (holder instanceof SentMsgHolder) {
            SentMsgHolder viewHolder = (SentMsgHolder) holder;
            if ("photo".equals(message.getMessage())) {
                viewHolder.binding.image.setVisibility(View.VISIBLE);
                viewHolder.binding.message.setVisibility(View.GONE);
                viewHolder.binding.mLinear.setVisibility(View.GONE);
                Glide.with(context)
                        .load(message.getImageUrl())
                        .placeholder(R.drawable.placeholder)
                        .into(viewHolder.binding.image);
            }
            viewHolder.binding.message.setText(message.getMessage());

            viewHolder.itemView.setOnLongClickListener(view -> {
                View dialogView = LayoutInflater.from(context).inflate(R.layout.delete_layout, null);
                DeleteLayoutBinding binding = DeleteLayoutBinding.bind(dialogView);
                AlertDialog dialog = new AlertDialog.Builder(context)
                        .setTitle("Delete Message")
                        .setView(binding.getRoot())
                        .create();

                binding.everyone.setOnClickListener(v -> {
                    message.setMessage("This message is removed.");
                    String messageId = message.getMessageId();
                    if (messageId != null) {
                        FirebaseDatabase.getInstance( RescuePetsRemoteRepository.firebaseRealtimeDatabaseUrl )
                                .getReference()
                                .child("Chats")
                                .child(senderRoom)
                                .child("Messages")
                                .child(messageId)
                                .setValue(message);
                        FirebaseDatabase.getInstance(RescuePetsRemoteRepository.firebaseRealtimeDatabaseUrl )
                                .getReference()
                                .child("Chats")
                                .child(receiverRoom)
                                .child("Messages")
                                .child(messageId)
                                .setValue(message);
                    }
                    dialog.dismiss();
                });

                binding.delete.setOnClickListener(v -> {
                    String messageId = message.getMessageId();
                    if (messageId != null) {
                        FirebaseDatabase.getInstance(RescuePetsRemoteRepository.firebaseRealtimeDatabaseUrl )
                                .getReference()
                                .child("Chats")
                                .child(senderRoom)
                                .child("Messages")
                                .child(messageId)
                                .setValue(null);
                    }
                    dialog.dismiss();
                });

                binding.cancel.setOnClickListener(v -> dialog.dismiss());

                dialog.show();
                return false;
            });
        } else if (holder instanceof ReceiveMsgHolder) {
            ReceiveMsgHolder viewHolder = (ReceiveMsgHolder) holder;
            if ("photo".equals(message.getMessage())) {
                viewHolder.binding.image.setVisibility(View.VISIBLE);
                viewHolder.binding.message.setVisibility(View.GONE);
                viewHolder.binding.mLinear.setVisibility(View.GONE);
                Glide.with(context)
                        .load(message.getImageUrl())
                        .placeholder(R.drawable.placeholder)
                        .into(viewHolder.binding.image);
            }
            viewHolder.binding.message.setText(message.getMessage());

            viewHolder.itemView.setOnLongClickListener(view -> {
                View dialogView = LayoutInflater.from(context).inflate(R.layout.delete_layout, null);
                DeleteLayoutBinding binding = DeleteLayoutBinding.bind(dialogView);
                AlertDialog dialog = new AlertDialog.Builder(context)
                        .setTitle("Delete Message")
                        .setView(binding.getRoot())
                        .create();

                binding.everyone.setOnClickListener(v -> {
                    message.setMessage("This message is removed.");
                    String messageId = message.getMessageId();
                    if (messageId != null) {
                        FirebaseDatabase.getInstance(RescuePetsRemoteRepository.firebaseRealtimeDatabaseUrl )
                                .getReference()
                                .child("Chats")
                                .child(senderRoom)
                                .child("Messages")
                                .child(messageId)
                                .setValue(message);
                        FirebaseDatabase.getInstance(RescuePetsRemoteRepository.firebaseRealtimeDatabaseUrl )
                                .getReference()
                                .child("Chats")
                                .child(receiverRoom)
                                .child("Messages")
                                .child(messageId)
                                .setValue(message);
                    }
                    dialog.dismiss();
                });

                binding.delete.setOnClickListener(v -> {
                    String messageId = message.getMessageId();
                    if (messageId != null) {
                        FirebaseDatabase.getInstance(RescuePetsRemoteRepository.firebaseRealtimeDatabaseUrl )
                                .getReference()
                                .child("Chats")
                                .child(senderRoom)
                                .child("Messages")
                                .child(messageId)
                                .setValue(null);
                    }
                    dialog.dismiss();
                });

                binding.cancel.setOnClickListener(v -> dialog.dismiss());

                dialog.show();
                return false;
            });
        }
    }

    @Override
    public int getItemCount() {
        return messages.size();
    }

    public class SentMsgHolder extends RecyclerView.ViewHolder {
        SendMsgBinding binding;

        public SentMsgHolder(View itemView) {
            super(itemView);
            binding = SendMsgBinding.bind(itemView);
        }
    }

    public class ReceiveMsgHolder extends RecyclerView.ViewHolder {
        ReceiveMsgBinding binding;

        public ReceiveMsgHolder(View itemView) {
            super(itemView);
            binding = ReceiveMsgBinding.bind(itemView);
        }
    }
}

File: Licenta/Anul 3/Semestrul 2/Android/Proiect/EmergencyApp/app/src/main/java/com/lifeSavers/emergencyapp/adapter/MessagesAdapter.kt
package com.lifeSavers.emergencyapp.adapter

import android.annotation.SuppressLint
import android.app.AlertDialog
import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.FirebaseDatabase
import com.lifeSavers.emergencyapp.R
import com.lifeSavers.emergencyapp.databinding.DeleteLayoutBinding
import com.lifeSavers.emergencyapp.databinding.ReceiveMsgBinding
import com.lifeSavers.emergencyapp.databinding.SendMsgBinding
import com.lifeSavers.emergencyapp.model.Message

class MessagesAdapter(
    var context: Context,
    messages: ArrayList<Message>?,
    senderRoom: String,
    receiverRoom: String
) : RecyclerView.Adapter<RecyclerView.ViewHolder?>() {
    private lateinit var messages: ArrayList<Message>
    private val itemSent = 1
    private val itemReceive = 2
    private var senderRoom: String
    private var receiverRoom: String

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return if (viewType == itemSent) {
            val view = LayoutInflater.from(context).inflate(R.layout.send_msg, parent, false)
            SentMsgHolder(view)
        } else {
            val view = LayoutInflater.from(context).inflate(R.layout.receive_msg, parent, false)
            SentMsgHolder(view)
            ReceiveMsgHolder(view)
        }
    }

    override fun getItemViewType(position: Int): Int {
        val message: Message = messages[position]
        return if (FirebaseAuth.getInstance().uid == message.senderId) {
            itemSent
        } else {
            itemReceive
        }
    }

    @SuppressLint("InflateParams")
    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val message = messages[position]
        if (holder.javaClass == SentMsgHolder::class.java) {
            val viewHolder = holder as SentMsgHolder
            if (message.message.equals("photo")) {
                viewHolder.binding.image.visibility = View.VISIBLE
                viewHolder.binding.message.visibility = View.GONE
                viewHolder.binding.mLinear.visibility = View.GONE
                Glide.with(context)
                    .load(message.imageUrl)
                    .placeholder(R.drawable.placeholder)
                    .into(viewHolder.binding.image)
            }
            viewHolder.binding.message.text = message.message
            viewHolder.itemView.setOnLongClickListener {
                val view: View = LayoutInflater.from(context)
                    .inflate(R.layout.delete_layout, null)

                val binding: DeleteLayoutBinding = DeleteLayoutBinding.bind(view)

                val dialog: AlertDialog = AlertDialog.Builder(context)
                    .setTitle("Delete Message")
                    .setView(binding.root)
                    .create()

                binding.everyone.setOnClickListener {
                    message.message = "This message is removed."
                    message.messageId?.let { it1 ->
                        FirebaseDatabase.getInstance("https://emergencyapp-3a6bd-default-rtdb.europe-west1.firebasedatabase.app/")
                            .reference.child("Chats")
                            .child(senderRoom)
                            .child("Messages")
                            .child(it1).setValue(message)
                    }
                    message.messageId?.let { it1 ->
                        FirebaseDatabase.getInstance("https://emergencyapp-3a6bd-default-rtdb.europe-west1.firebasedatabase.app/")
                            .reference.child("Chats")
                            .child(receiverRoom)
                            .child("Messages")
                            .child(it1).setValue(message)
                    }
                    dialog.dismiss()
                }
                binding.delete.setOnClickListener {
                    message.messageId?.let { it1 ->
                        FirebaseDatabase.getInstance("https://emergencyapp-3a6bd-default-rtdb.europe-west1.firebasedatabase.app/")
                            .reference.child("Chats")
                            .child(senderRoom)
                            .child("Messages")
                            .child(it1).setValue(null)
                    }
                    dialog.dismiss()
                }
                binding.cancel.setOnClickListener { dialog.dismiss() }
                dialog.show()
                false
            }
        } else {
            val viewHolder = holder as ReceiveMsgHolder
            if (message.message.equals("photo")) {
                viewHolder.binding.image.visibility = View.VISIBLE
                viewHolder.binding.message.visibility = View.GONE
                viewHolder.binding.mLinear.visibility = View.GONE
                Glide.with(context)
                    .load(message.imageUrl)
                    .placeholder(R.drawable.placeholder)
                    .into(viewHolder.binding.image)
            }
            viewHolder.binding.message.text = message.message
            viewHolder.itemView.setOnLongClickListener {
                val view: View = LayoutInflater.from(context)
                    .inflate(R.layout.delete_layout, null)

                val binding: DeleteLayoutBinding = DeleteLayoutBinding.bind(view)

                val dialog: AlertDialog = AlertDialog.Builder(context)
                    .setTitle("Delete Message")
                    .setView(binding.root)
                    .create()

                binding.everyone.setOnClickListener {
                    message.message = "This message is removed."
                    message.messageId?.let { it1 ->
                        FirebaseDatabase.getInstance("https://emergencyapp-3a6bd-default-rtdb.europe-west1.firebasedatabase.app/")
                            .reference.child("Chats")
                            .child(senderRoom)
                            .child("Messages")
                            .child(it1).setValue(message)
                    }
                    message.messageId.let { it1 ->
                        FirebaseDatabase.getInstance("https://emergencyapp-3a6bd-default-rtdb.europe-west1.firebasedatabase.app/")
                            .reference.child("Chats")
                            .child(receiverRoom)
                            .child("Messages")
                            .child(it1!!).setValue(message)
                    }
                    dialog.dismiss()
                }
                binding.delete.setOnClickListener {
                    message.messageId.let { it1 ->
                        FirebaseDatabase.getInstance("https://emergencyapp-3a6bd-default-rtdb.europe-west1.firebasedatabase.app/")
                            .reference.child("Chats")
                            .child(senderRoom)
                            .child("Messages")
                            .child(it1!!).setValue(null)
                    }
                    dialog.dismiss()
                }
                binding.cancel.setOnClickListener { dialog.dismiss() }
                dialog.show()
                false
            }
        }
    }

    override fun getItemCount(): Int {
        return messages.size
    }

    inner class SentMsgHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var binding: SendMsgBinding

        init {
            binding = SendMsgBinding.bind(itemView)
        }
    }

    inner class ReceiveMsgHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var binding: ReceiveMsgBinding

        init {
            binding = ReceiveMsgBinding.bind(itemView)
        }
    }

    init {
        if (messages != null) {
            this.messages = messages
        }
        this.senderRoom = senderRoom
        this.receiverRoom = receiverRoom
    }

}